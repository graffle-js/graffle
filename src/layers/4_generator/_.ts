export * from './config.js'
export * from './generate.js'
