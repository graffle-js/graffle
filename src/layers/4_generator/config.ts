import { buildClientSchema, printSchema } from 'graphql'
import { buildSchema, type GraphQLObjectType, type GraphQLSchema } from 'graphql'
import fs from 'node:fs/promises'
import * as Path from 'node:path'
import { introspectionQuery } from '../../cli/_helpers.js'
import { getTypeMapByKind, type TypeMapByKind } from '../../lib/graphql.js'
import { omitUndefinedKeys } from '../../lib/prelude.js'
import { fileExists } from './helpers/fs.js'

export interface Input {
  sourceSchema: {
    type: 'sdl'
    /**
     * Defaults to the source directory if set, otherwise the current working directory.
     */
    dirPath?: string
  } | {
    type: 'url'
    url: URL
  }
  outputDirPath?: string
  name?: string
  // code?: Omit<GenerateCodeInput, 'schemaSource' | 'sourceDirPath' | 'options'>
  defaultSchemaUrl?: boolean | URL
  /**
   * Defaults to the current working directory.
   */
  sourceDirPath?: string
  sourceCustomScalarCodecsFilePath?: string
  schemaPath?: string
  /**
   * Override import paths to graffle package within the generated code.
   * Used by Graffle test suite to have generated clients point to source
   * code. Probably not useful to you.
   */
  libraryPaths?: InputLibraryPaths
  errorTypeNamePattern?: RegExp
  /**
   * Should custom scalars definitions be imported into the generated output?
   */
  customScalars?: boolean
  format?: boolean
  TSDoc?: {
    noDocPolicy?: 'message' | 'ignore'
  }
}

interface InputLibraryPaths {
  client?: string
  schema?: string
  scalars?: string
  utilitiesForGenerated?: string
}

export interface Config {
  name: string
  paths: {
    project: {
      outputs: {
        root: string
        modules: string
      }
      inputs: {
        root: string
        customScalarCodecs: string
      }
    }
    imports: {
      customScalarCodecs: string
      grafflePackage: Required<InputLibraryPaths>
    }
  }
  schema: {
    instance: GraphQLSchema
    typeMapByKind: TypeMapByKind
    error: {
      objects: GraphQLObjectType[]
      enabled: boolean
    }
  }
  options: {
    defaultSchemaUrl: URL | null
    format: boolean
    errorTypeNamePattern: RegExp | null
    customScalars: boolean
    TSDoc: {
      noDocPolicy: 'message' | 'ignore'
    }
  }
}

export const defaultName = `default`

export const defaultLibraryPaths = {
  client: `graffle/client`,
  scalars: `graffle/schema/scalars`,
  schema: `graffle/schema`,
  utilitiesForGenerated: `graffle/utilities-for-generated`,
}

export const createConfig = async (input: Input): Promise<Config> => {
  // dprint-ignore
  const defaultSchemaUrl =
    typeof input.defaultSchemaUrl === `boolean`
      ? input.sourceSchema.type === `url`
        ? input.sourceSchema.url
        : null
      : input.defaultSchemaUrl??null

  const formattingEnabled = input.format ?? true

  const inputPathDirRoot = input.sourceDirPath ?? process.cwd()

  const outputDirPathRoot = input.outputDirPath ?? Path.join(process.cwd(), `./graffle`)

  const outputDirPathModules = Path.join(outputDirPathRoot, `/modules`)

  const inputPathCustomScalarCodecs = input.sourceCustomScalarCodecsFilePath
    ?? Path.join(inputPathDirRoot, `customScalarCodecs.ts`)

  const customScalarCodecsPathExists = await fileExists(inputPathCustomScalarCodecs)

  const customScalarCodecsImportPath = Path.relative(
    outputDirPathModules,
    inputPathCustomScalarCodecs.replace(/\.ts$/, `.js`),
  )

  const errorTypeNamePattern = input.errorTypeNamePattern ?? null

  const sourceSchema = await resolveSourceSchema(input)

  const schema = buildSchema(sourceSchema)
  const typeMapByKind = getTypeMapByKind(schema)
  const errorObjects = errorTypeNamePattern
    ? Object.values(typeMapByKind.GraphQLObjectType).filter(_ => _.name.match(errorTypeNamePattern))
    : []

  // const rootTypes = {
  //   Query: typeMapByKind.GraphQLRootType.find(_ => _.name === `Query`) ?? null,
  //   Mutation: typeMapByKind.GraphQLRootType.find(_ => _.name === `Mutation`) ?? null,
  //   Subscription: typeMapByKind.GraphQLRootType.find(_ => _.name === `Subscription`) ?? null,
  // }

  return {
    name: input.name ?? defaultName,
    paths: {
      project: {
        outputs: {
          root: outputDirPathRoot,
          modules: outputDirPathModules,
        },
        inputs: {
          root: inputPathDirRoot,
          customScalarCodecs: inputPathCustomScalarCodecs,
        },
      },
      imports: {
        customScalarCodecs: customScalarCodecsImportPath,
        grafflePackage: {
          ...defaultLibraryPaths,
          ...omitUndefinedKeys(input.libraryPaths ?? {}),
        },
      },
    },
    schema: {
      instance: schema,
      typeMapByKind,
      error: {
        enabled: Boolean(errorTypeNamePattern),
        objects: errorObjects,
      },
    },
    options: {
      defaultSchemaUrl,
      format: formattingEnabled,
      errorTypeNamePattern,
      customScalars: customScalarCodecsPathExists,
      TSDoc: {
        noDocPolicy: input.TSDoc?.noDocPolicy ?? `ignore`,
      },
    },
  }
}

const resolveSourceSchema = async (input: Input) => {
  if (input.sourceSchema.type === `sdl`) {
    const sourceDirPath = input.sourceSchema.dirPath ?? input.sourceDirPath ?? process.cwd()
    const schemaPath = input.schemaPath ?? Path.join(sourceDirPath, `schema.graphql`)
    const sdl = await fs.readFile(schemaPath, `utf8`)
    return sdl
  } else {
    const data = await introspectionQuery(input.sourceSchema.url)
    const schema = buildClientSchema(data)
    const sdl = printSchema(schema)
    return sdl
  }
}
