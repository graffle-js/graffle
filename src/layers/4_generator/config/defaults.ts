export const defaultName = `default`

export const defaultLibraryPaths = {
  client: `graffle/client`,
  scalars: `graffle/schema/scalars`,
  schema: `graffle/schema`,
  utilitiesForGenerated: `graffle/utilities-for-generated`,
}
