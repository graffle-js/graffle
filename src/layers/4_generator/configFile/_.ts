export * from './builder.js'
export * from './loader.js'
