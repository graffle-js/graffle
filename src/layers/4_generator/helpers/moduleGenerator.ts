import {
  createCodeGenerator,
  type ModuleGeneratorRunner,
  type ModuleGeneratorRunnerImplementation,
} from './moduleGeneratorRunner.js'

export type FactoryModuleGenerator = <$Name extends string>(
  /**
   * The name of the file that will be written to disk.
   */
  name: $Name,
  runnerImplementation: ModuleGeneratorRunnerImplementation,
) => ModuleGenerator<$Name>

export interface ModuleGenerator<$Name extends string = string> {
  /**
   * The name of the file that will be written to disk.
   */
  name: $Name
  generate: ModuleGeneratorRunner
}

export interface GeneratedModule<$Name extends string = string> {
  name: $Name
  content: string
}

export const createModuleGenerator: FactoryModuleGenerator = (name, runnerImplementation) => {
  const runner = createCodeGenerator(runnerImplementation)

  const generate: ModuleGeneratorRunner = (config) => {
    const content = runner({ config })
    return {
      content,
      name,
    }
  }

  return {
    name,
    generate,
  }
}
