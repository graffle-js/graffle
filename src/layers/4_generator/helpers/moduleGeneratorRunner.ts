import type { Config } from '../config.js'
import type { GeneratedModule } from './moduleGenerator.js'

type FactoryModuleGeneratorRunner = <$CustomInput extends object = {}>(
  runnerImplementation: ModuleGeneratorRunnerImplementation<$CustomInput>,
) => CodeGenerator<$CustomInput>

export type ModuleGeneratorRunner = (config: Config) => GeneratedModule

export type CodeGenerator<$CustomInput extends object = {}> = (input: $CustomInput & BaseInput) => Code

export type ModuleGeneratorRunnerImplementation<$CustomInput extends object = {}> = (
  input: $CustomInput & BaseInputInternal,
) => void

interface BaseInputInternal extends BaseInput {
  code: LinesOfGeneratedCode
}

interface BaseInput {
  config: Config
}

type LinesOfGeneratedCode = (Code | null)[]

type Code = string

export const createModuleGeneratorRunner: FactoryModuleGeneratorRunner = (runnerImplementation) => {
  return (input) => {
    const code: LinesOfGeneratedCode = []
    runnerImplementation({ ...input, code })
    return code.filter(_ => _ !== null).map(_ => _.trim()).join(`\n`)
  }
}
