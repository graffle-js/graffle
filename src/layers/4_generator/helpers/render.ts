import { Code } from '../../../lib/Code.js'
import { Grafaid } from '../../../lib/grafaid/__.js'
import { getNodeDisplayName } from '../../../lib/grafaid/graphql.js'
import { borderThickFullWidth, borderThinFullWidth, centerTo } from '../../../lib/text.js'
import type { Config } from '../config.js'

export const title1 = (title: string, subTitle?: string) => {
  const titleDecorated = `
    //
    //
    //
    //
    //
    //
    // ${borderThickFullWidth}
    // ${centerTo(borderThickFullWidth, title)}${subTitle ? `\n// ${centerTo(borderThickFullWidth, subTitle)}` : ``}
    // ${borderThickFullWidth}
    //
    //
    //
    //
    //
    //
  `
  return titleDecorated
}

export const typeTitle2 = (category: string) => (type: Grafaid.Schema.NamedTypes) => {
  const typeKind = Grafaid.getTypeNameAndKind(type)
  const nameOrKind = typeKind.kind === `Scalar` ? typeKind.name : typeKind.kind
  const typeLabel = nameOrKind
  const title = `
    //
    //
    //
    //
    // ${category.toUpperCase()}
    // ${typeLabel.toUpperCase()}
    // ${borderThinFullWidth}
    // ${centerTo(borderThinFullWidth, type.name)}
    // ${borderThinFullWidth}
    //
    //
  `.trim()

  return title
}

export const typeTitle2SelectionSet = typeTitle2(`GRAPHQL SELECTION SET`)

export const typeTitle = (config: Config, typeKind: Grafaid.Schema.NamedTypeKind) => {
  const hasItems = config.schema.typeMapByKind[`GraphQL${typeKind}Type`].length > 0
  const title = `${typeKind} Types`
  const titleDecorated = `// ${title}\n// ${`-`.repeat(title.length)}\n`
  if (hasItems) {
    return titleDecorated
  } else {
    return `${titleDecorated}\n// -- None --\n`
  }
}

const defaultDescription = (node: Grafaid.Schema.DescribableTypes) =>
  `There is no documentation for this ${getNodeDisplayName(node)}.`

export const renderDocumentation = (config: Config, node: Grafaid.Schema.DescribableTypes) => {
  return Code.TSDoc(getDocumentation(config, node))
}
export const getDocumentation = (config: Config, node: Grafaid.Schema.DescribableTypes) => {
  const generalDescription = node.description
    ?? (config.options.TSDoc.noDocPolicy === `message` ? defaultDescription(node) : null)

  const deprecationDescription = Grafaid.Schema.isDeprecatableNode(node) && node.deprecationReason
    ? `@deprecated ${node.deprecationReason}`
    : null

  const enumMemberDescriptions: string[] = Grafaid.Schema.isEnumType(node)
    ? node
      .getValues()
      .map((_) => {
        const deprecationDescription = _.deprecationReason
          ? `(DEPRECATED: ${_.deprecationReason})`
          : null
        const generalDescription = _.description
          ? _.description
          : config.options.TSDoc.noDocPolicy === `message`
          ? `Missing description.`
          : null
        if (!generalDescription && !deprecationDescription) return null
        const content = [generalDescription, deprecationDescription]
          .filter((_) => _ !== null)
          .join(` `)
        return [_, content] as const
      })
      .filter((_): _ is [Grafaid.Schema.EnumValue, string] => _ !== null)
      .map(([node, description]) => {
        const content = `"${node.name}" - ${description}`
        return content
      })
    : []
  const enumMemberDescription = enumMemberDescriptions.length > 0
    ? `Members\n${enumMemberDescriptions.join(`\n`)}`
    : null
  if (!enumMemberDescription && !generalDescription && !deprecationDescription) {
    return null
  }
  const content = [
    generalDescription,
    enumMemberDescription,
    deprecationDescription,
  ]
    .filter((_) => _ !== null)
    .join(`\n\n`)
  return content
}

/**
 * Render the type name. Generally just a passthrough but
 * this guards against GraphQL type or property names that
 * would be illegal in TypeScript such as `namespace` or `interface`.
 */
export const renderName = (type: Grafaid.Schema.NamedTypes | Grafaid.Schema.Field<any, any>) => {
  if (Code.reservedTypeScriptInterfaceNames.includes(type.name as any)) {
    return `$${type.name}`
  }
  return type.name
}
