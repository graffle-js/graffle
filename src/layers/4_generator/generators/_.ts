import { createModuleGenerator } from '../helpers/moduleGenerator.js'
import { ModuleGeneratorClient } from './Client.js'
import { ModuleGeneratorError } from './Error.js'
import { ModuleGeneratorSelect } from './Select.js'

export const ModuleGenerator_ = createModuleGenerator(
  `_`,
  ({ code }) => {
    code(
      `// We import the global module for good measure although it is not clear it is always needed.`,
      `// It at least helps with Twoslash wherein without this import here Twoslash will not include the global module.`,
      `// In real TypeScript projects it seems the global module is included automatically. But there could be certain tsconfig`,
      `// setups where this still indeed does help.`,
      `import './modules/Global.js'`,
      ``,
      `export { Select } from './modules/${ModuleGeneratorSelect.name}.js'`,
      `export { isError } from './modules/${ModuleGeneratorError.name}.js'`,
      `export { create } from './modules/${ModuleGeneratorClient.name}.js'`,
    )

    return code
  },
)
