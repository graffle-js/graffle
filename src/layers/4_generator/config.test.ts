import _ from 'json-bigint'
import { expect } from 'vitest'
import { test } from '../../../tests/_/helpers.js'
import { createConfig } from './config.js'

test(`can load schema from custom path`, async () => {
  const customPathFile = `./tests/_/fixtures/custom.graphql`
  const config = await createConfig({ sourceSchema: { type: `sdl`, dirOrFilePath: customPathFile } })
  const field = config.schema.instance.getQueryType()?.getFields()[`customNamedSchemaFile`]
  expect(config.paths.project.inputs.schema).toEqual(customPathFile)
  expect(config.schema.sdl).toMatchSnapshot()
  expect(field).toBeDefined()
})

test(`can load schema from custom dir using default file name`, async () => {
  const customPathDir = `tests/_/fixtures`
  const config = await createConfig({ sourceSchema: { type: `sdl`, dirOrFilePath: customPathDir } })
  const field = config.schema.instance.getQueryType()?.getFields()[`defaultNamedSchemaFile`]
  expect(config.paths.project.inputs.schema).toEqual(customPathDir + `/schema.graphql`)
  expect(config.schema.sdl).toMatchSnapshot()
  expect(field).toBeDefined()
})

test(`can introspect schema from url`, async ({ pokemonService }) => {
  const config = await createConfig({ sourceSchema: { type: `url`, url: pokemonService.url } })
  expect(config.paths.project.inputs.schema).toEqual(null)
  expect(config.schema.sdl).toMatchSnapshot()
})
