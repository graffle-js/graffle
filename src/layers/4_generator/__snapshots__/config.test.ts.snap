// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`can introspect schema from url 1`] = `
"interface Being {
  id: Int
  name: String
}

input DateFilter {
  gte: Float
  lte: Float
}

type Mutation {
  addPokemon(attack: Int, defense: Int, hp: Int, name: String!, type: PokemonType!): Pokemon
}

type Patron implements Being {
  id: Int
  money: Int
  name: String
}

type Pokemon implements Being {
  attack: Int
  birthday: Int
  defense: Int
  hp: Int
  id: Int
  name: String
  trainer: Trainer
  type: PokemonType
}

input PokemonFilter {
  birthday: DateFilter
  name: StringFilter
}

enum PokemonType {
  electric
  fire
  grass
  water
}

type Query {
  beings: [Being!]!
  pokemon: [Pokemon!]
  pokemonByName(name: String!): [Pokemon!]
  pokemons(filter: PokemonFilter): [Pokemon!]
  trainerByName(name: String!): Trainer
  trainers: [Trainer!]
}

input StringFilter {
  contains: String
  in: [String!]
}

type Trainer implements Being {
  class: TrainerClass
  fans: [Patron!]
  id: Int
  name: String
  pokemon: [Pokemon!]
}

enum TrainerClass {
  bugCatcher
  camper
  picnicker
  psychic
  psychicMedium
  psychicYoungster
  sailor
  superNerd
  tamer
  teamRocketGrunt
  triathlete
  youngster
  youth
}"
`;

exports[`can load schema from custom dir using default file name 1`] = `
"type Query {
	defaultNamedSchemaFile: Boolean
}
"
`;

exports[`can load schema from custom path 1`] = `
"type Query {
	customNamedSchemaFile: Boolean
}
"
`;
